cpp/include/typedb/concept/thing/thing.hpp

```diff
@@ -0,0 +1,58 @@
+
+/*
+ * Copyright (C) 2022 Vaticle
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#pragma once
+
+#include "typedb/concept/concept.hpp"
+
+namespace TypeDB {
+
+class Thing : public Concept { // TODO: Make abstract?
+   public:
+    std::string getIID();
+    std::unique_ptr<ThingType> getType();
+    bool isInferred();
+    BoolFuture isDeleted(Transaction& transaction);
+    VoidFuture drop(Transaction& transaction);
+
+    [[nodiscard]] VoidFuture setHas(Transaction& transaction, Attribute* attribute);
+    [[nodiscard]] VoidFuture unsetHas(Transaction& transaction, Attribute* attribute);
+
+    ConceptIterable<Attribute> getHas(Transaction& transaction);
+    #pragma message("UX messages for variadic args when I use vectors instead")
+    ConceptIterable<Attribute> getHas(Transaction& transaction, const AttributeType* attribute);
+    ConceptIterable<Attribute> getHas(Transaction& transaction, const std::vector<const AttributeType*>& attributeTypes);
+    ConceptIterable<Attribute> getHas(Transaction& transaction, const std::vector<Annotation>& annotations);
```

Comment:
@dmitrii-ubskii 
I need design input / an opinion on my choice of using `const std::vector<ConceptType*>&` for those arguments.

Particularly in the light of how we do things with Annotations, I'm wondering whether they should be
`const std::vector<std::unique_ptr<ConceptType>>&`, to avoid ownership issues. I'll need to create a new vector for each native call, but that feels like an acceptable cost.

Please give thoughts.

Reply:
In general, when a reference to a Rust value is accepted as a reference, it is read-only, i.e. the memory will not be freed after the call resolves. Accepting `T const *` rather than `std::unique_ptr<T>` reflects that intent.
For the sake of the UX however, we can overload each place that accepts a collection of raw `T*` pointers to also accept a collection of `std::unique_ptr`s, though I'm not sure how useful that'd be.

Reply:
After a discord decision, we've decided to add a  `const std::vector<std::unique_ptr<ConceptType>>&` variant